File: JavaAlgorithm/src/main/java/algo/tzashinorpu/SpecialSubject/BinaryTree/convertBST_538.java
package algo.tzashinorpu.SpecialSubject.BinaryTree;

import algo.tzashinorpu.SpecialSubject.TreeNode;

public class convertBST_538 {
    private int sum = 0;

    public TreeNode convertBST(TreeNode root) {
        if (root == null) return null;
        convertBST(root.right);
        root.val += sum;
        sum = root.val;
        convertBST(root.left);
        return root;
    }
}

File: Clion/ZeroTrac/Tree/convertBST_538.cpp
//
// Created by TzashiNorpu on 11/3/2022.
//
#include "tree.h"

using namespace Tree;

TreeNode *Solution::convertBST(TreeNode *root) {
  int sum = 0;
  function<TreeNode *(TreeNode *)> helper = [&](TreeNode *node) -> TreeNode * {
      if (!node) return node;
      helper(node->right);
      node->val += sum;
      sum = node->val;
      helper(node->left);
      return node;
  };
  return helper(root);
}